### ISSUE TITLE
Feature request: some API for parsing from a string/buffer

### ISSUE THREAD

**Issue by nfischer** [freitagbr/crystal-uvarint#1]
It would be really great to have a way to build a varint from an array of bytes where we may not know how long the Varint itself is. Imagine you have:

```
[1000 0001 , 1000 0010 , 0000 0001 , 1000 0001 , 0000 0001, 1000 0010]
````

It would be nice to have a constructor for Varints like [Go's implementation](https://github.com/golang/go/blob/master/src/encoding/binary/varint.go#L52):

```crystal
value1, numBytesRead = Varint(bytes)
puts value1 # => 16641 (if my math is right)
puts numBytesRead # => 3

# Now we can get the second Varint from the array!
newBytes = getArrayStartingAt(bytes, numBytesRead) # => [1000 0001, 0000 0001, 1000 0010]

value2, numBytesRead = Varint(newBytes)
puts value2 # => 129 (if my math is right)
puts numBytesRead # => 2

lastByteArray = getArrayStartingAt(newBytes, numBytesRead) # => [1000 0010]
value3, numBytesRead = Varint(lastByteArray) # Throws an exception! We can't parse this
```

As with the final example, if we can't parse a VarInt from the byte array, then we should just throw an exception.

This could either be implemented as a class constructor or a module-level method that returns a Varint.

**Comment by freitagbr** [comment]
Suggestion: reimplement as a `struct` (a `Varint` should be immutable). Allow instantiation from  an `Int`, and `Array`, or `Bytes`. Add `encoded` and `decoded` methods that return the `UInt64` and `Bytes` representations, respectively.

**Comment by nfischer** [comment]
If we're returning `UInt64` from an instance, how about we override [`.to_i` and its similar functions](https://github.com/crystal-lang/crystal/blob/master/src/string.cr)? `.to_u64` would be the one which actually returns the `UInt64`.

This might be a good opportunity to use templates for metaprogramming. That keeps the code base simple to maintain, but very flexible and extensible.

Also, it makes sense to provide a `.to_s(base : Int32)`. Then Multihash can use `.to_s(16)` for serializing. If we want to expose the Bytes, we can add a method for that specifically.

**Comment by freitagbr** [comment]
This should be closed by #2. For multihash, we can use the method `.hexstring` for the desired behavior.

**Comment by nfischer** [comment]
@freitagbr this wasn't fixed in #2 (see the title & description). #2 addressed some of the extra features mentioned in comment 1 & 2. Actually, #4 is a duplicate of this.

**Comment by freitagbr** [comment]
Oh, you're right. I will mark #4 as a duplicate. PR #5 should fix the remaining issues here.

**Comment by freitagbr** [comment]
This should be closed by #5.
